consensus_pileup.c

```diff
@@ -0,0 +1,606 @@
+            : 0;
+        break;
+
+    case BAM_CHARD_CLIP:
+        p->eof = 1;
+        break;
+
+    default:
+        break;
+    }
+
+    return 1;
+}
+
+/*
+ * Loops through a set of supplied ranges producing columns of data.
+ * When found, it calls func with clientdata as a callback. Func should
+ * return 0 for success and non-zero for failure. seq_init() is called
+ * on each new entry before we start processing it. It should return 0 or 1
+ * to indicate reject or accept status (eg to filter unmapped data).
+ * If seq_init() returns -1 we abort the pileup_loop with an error.
+ * seq_init may be NULL.
+ *
+ * Returns 0 on success
+ *        -1 on failure
+ */
+int pileup_loop(samFile *fp,
+                sam_hdr_t *h,
+                int (*seq_fetch)(void *client_data,
+                                 samFile *fp,
+                                 sam_hdr_t *h,
+                                 bam1_t *b),
+                int (*seq_init)(void *client_data,
+                                samFile *fp,
+                                sam_hdr_t *h,
+                                pileup_t *p),
+                int (*seq_add)(void *client_data,
+                               samFile *fp,
+                               sam_hdr_t *h,
+                               pileup_t *p,
+                               int depth,
+                               int pos,
+                               int nth,
+                               int is_insert),
+                void *client_data) {
+    int ret = -1;
+    pileup_t *phead = NULL, *p, *pfree = NULL, *last, *next, *ptail = NULL;
+    pileup_t *pnew = NULL;
+    int is_insert, nth = 0;
+    int col = 0, r;
+    int last_ref = -1;
+
+    /* FIXME: allow for start/stop boundaries rather than consuming all data */
+
+    if (NULL == (pnew = calloc(1, sizeof(*p))))
+        return -1;
+
+    do {
+        bam1_t *b;
+        int pos, last_in_contig;
+
+        //r = scram_next_seq(fp, &pnew->b);
+        r = seq_fetch(client_data, fp, h, &pnew->b);
+        //r = sam_read1(fp, h, &pnew->b); // FIXME: use readaln
+        if (r < -1) {
+            fprintf(stderr, "bam_next_seq() failure.\n");
+            return -1;
+        }
+
+        b = &pnew->b;
+
+        /* Force realloc */
+        //fp->bs = NULL;
+        //fp->bs_size = 0;
+
+        //r = samread(fp, pnew->b);
+        if (r >= 0) {
+            if (b->core.flag & BAM_FUNMAP)
+                continue;
+
+            if (b->core.tid == -1) {
+                /* Another indicator for unmapped */
+                continue;
+            } else if (b->core.tid == last_ref) {
+                pos = b->core.pos+1;
+                //printf("New seq at pos %d @ %d %s\n", pos, b->core.tid,
+                //       bam_name(b));
+                last_in_contig = 0;
+            } else {
+                //printf("New ctg at pos %d @ %d\n",b->core.pos+1,b->core.tid);
+                pos = (b->core.pos > col ? b->core.pos : col)+1;
```

Comment:
A reasonable comment on `last_in_contig`.  It's set when we change contigs, so isn't so much that _this_ read is the last but that the previous one was, so a change of name is OK.

As for `col+1`, oddly it makes no difference as when `last_in_contig` is set we have `pos++` later on for this case and both col and pos march up until we hit the end of the contig.   Hence any value of pos is sufficient providing it's above col.  The simpler definition you suggest is therefore the obvious one.  (I'm struggling to remember precisely why that increment is there, but it's needed.  That probably means I should have commented that bit!)

`HTS_POS_MAX` is a bad idea though.  Yes it does work, but only because it's defined to be 0x7fffffff80000000 rather than 0x7fffffffffffffff, which luckily avoids the overflow when we increment pos beyond `HTS_POS_MAX`.

Reply:
If you used `HTS_POS_MAX` then you could get rid of the line where it increments `pos` as it will already be bigger than any sensible value for `col`.  So I don't think `pos` could overflow.

Reply:
I did try this and it fails, but I've now figured out why.  (There's more than one place it needed changing in.)  This is indeed a better solution.  Thanks